File: java/amp-bus/src/main/java/amp/bus/DefaultEnvelopeBus.java
package amp.bus;


import java.util.LinkedList;
import java.util.List;

import cmf.bus.Envelope;
import cmf.bus.IEnvelopeBus;
import cmf.bus.IRegistration;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import amp.bus.EnvelopeContext.Directions;


public class DefaultEnvelopeBus implements IEnvelopeBus, IEnvelopeReceivedCallback {

	protected final Logger log = LoggerFactory.getLogger(this.getClass().getCanonicalName());
	
    protected List<IEnvelopeProcessor> _inboundProcessors = new LinkedList<IEnvelopeProcessor>();
    protected List<IEnvelopeProcessor> _outboundProcessors = new LinkedList<IEnvelopeProcessor>();
    protected ITransportProvider _transportProvider;

    
    public void setInboundProcessors(List<IEnvelopeProcessor> inboundChain) {
    	_inboundProcessors = inboundChain;
    }
    
    public void setOutboundProcessors(List<IEnvelopeProcessor> outboundChain) {
    	_outboundProcessors = outboundChain;
    }
    
    
    public DefaultEnvelopeBus(ITransportProvider transportProvider) {
        _transportProvider = transportProvider;

        initialize();
    }

    public DefaultEnvelopeBus(
		ITransportProvider transportProvider, 
		List<IEnvelopeProcessor> inboundChain,
        List<IEnvelopeProcessor> outboundChain)
    {
        this._transportProvider = transportProvider;
        this._inboundProcessors = inboundChain;
        this._outboundProcessors = outboundChain;

        initialize();
    }

    
    @Override
    public void dispose() {
        _transportProvider.dispose();

        for (IEnvelopeProcessor p : _inboundProcessors) {
            try {
                p.dispose();
            } catch (Exception ex) {
            }
        }

        for (IEnvelopeProcessor p : _outboundProcessors) {
            try {
                p.dispose();
            } catch (Exception ex) {
            }
        }
    }

    public void processEnvelope(
    		final EnvelopeContext context, 
    		final List<IEnvelopeProcessor> processingChain, 
    		final IContinuationCallback onComplete) throws Exception
    {
    	log.debug("Enter processEnvelope");
    	
    	// if the chain is null or empty, complete processing
    	if ( (null == processingChain) || (0 == processingChain.size()) ) {
    		log.debug("envelope processing complete");
    		onComplete.continueProcessing();
    		return;
    	}
    	
    	// get the first processor
    	IEnvelopeProcessor processor = processingChain.get(0);
    	
    	// create a processing chain that no longer contains this processor
    	final List<IEnvelopeProcessor> newChain = processingChain.subList(1, processingChain.size());
    	
    	// let it process the envelope and pass its "next" processor: a method that
    	// recursively calls this function with the current processor removed
    	processor.processEnvelope(context, new IContinuationCallback() {

			@Override
			public void continueProcessing() throws Exception {
				processEnvelope(context, newChain, onComplete);
			}
    		
    	});
    	
    	log.debug("Leave processEnvelope");
    }
    
    @Override
    public void register(final IRegistration registration) throws Exception {
        log.debug("Enter register");
        if (registration == null) {
            throw new IllegalArgumentException("Cannot register with a null registration");
        }

        _transportProvider.register(registration);
        log.debug("Leave register");
    }

    @Override
    public void send(final Envelope envelope) throws Exception {
    	
        log.debug("Enter send");
        
        // check for null envelopes
        if (envelope == null) {
            throw new IllegalArgumentException("Cannot send a null envelope");
        }

        // create an envelope context that will be passed among the processors
        final EnvelopeContext context = new EnvelopeContext(Directions.Out, envelope);
        
        // send the envelope through the outbound processing chain
        this.processEnvelope(context, _outboundProcessors, new IContinuationCallback() {

			@Override
			public void continueProcessing() throws Exception {
				_transportProvider.send(context.getEnvelope());
				log.debug("Outgoing headers: {}", new EnvelopeHelper(context.getEnvelope()).flatten());
			}
        	
        });
        
        log.debug("Leave send");
    }

    @Override
    public void unregister(IRegistration registration) throws Exception {
        log.debug("Enter unregister");

        if (registration == null) {
            throw new IllegalArgumentException("Cannot unregister with a null registration");
        }
        _transportProvider.unregister(registration);

        log.debug("Leave unregister");
    }

    @Override
	public void handleReceive(final IEnvelopeDispatcher dispatcher) {
    	
    	log.debug("Got an envelope dispatcher from the transport provider");
    	final EnvelopeContext context = new EnvelopeContext(Directions.In, dispatcher.getEnvelope());

        try {
            // send the envelope through the inbound processing chain
            this.processEnvelope(context, _inboundProcessors, new IContinuationCallback() {

                @Override
                public void continueProcessing() throws Exception {
                    // the dispatcher encapsulates the logic of giving the envelope to handlers
                    dispatcher.dispatch(context.getEnvelope());

                    log.debug("Dispatched envelope");
                }  
            });
        } catch (Exception ex) {
        	
            log.warn("Failed to dispatch envelope; raising EnvelopeFailed event: {}", ex);
            dispatcher.dispatchFailed(context.getEnvelope(), ex);
        }
	}

    
    void initialize() {

    	// add a handler to the transport provider's envelope received event
        _transportProvider.onEnvelopeReceived(this);

        log.debug("Initialized");
    }
}

File: dotnet/source/amp.bus/DefaultEnvelopeBus.cs
using System;
using System.Collections.Generic;
using System.Linq;

using Common.Logging;

using cmf.bus;


namespace amp.bus
{
    public class DefaultEnvelopeBus : IEnvelopeBus
    {
        protected ITransportProvider _txProvider;
        protected ILog _log;


        public IDictionary<int, IEnvelopeProcessor> InboundChain { get; set; }
        public IDictionary<int, IEnvelopeProcessor> OutboundChain { get; set; }


        public DefaultEnvelopeBus(ITransportProvider transportProvider)
        {
            _txProvider = transportProvider;
            _txProvider.OnEnvelopeReceived += this.Handle_Dispatcher;

            _log = LogManager.GetLogger(this.GetType());
        }


        public virtual void Send(Envelope env)
        {
            _log.Debug("Enter Send");

            // guard clause
            if (null == env) { throw new ArgumentNullException("Cannot send a null envelope"); }
            
            // create a context
            EnvelopeContext context = new EnvelopeContext(EnvelopeContext.Directions.Out, env);

            // send the envelope through the outbound chain
            this.ProcessEnvelope(context, this.OutboundChain.Sort(), () =>
            {
                // send the envelope to the transport provider
                _txProvider.Send(context.Envelope);

                // log the headers of the outgoing envelope
                _log.Debug(string.Format("Outgoing headers: {0}", context.Envelope.Headers.Flatten()));
            });

            _log.Debug("Leave Send");
        }

        public virtual void Register(IRegistration registration)
        {
            _log.Debug("Enter Register");

            // guard clause
            if (null == registration) { throw new ArgumentNullException("Cannot register a null registration"); }
            _txProvider.Register(registration);

            _log.Debug("Leave Register");
        }

        public virtual void Unregister(IRegistration registration)
        {
            _log.Debug("Enter Unregister");

            if (null == registration) { throw new ArgumentNullException("Cannot unregister a null registration"); }
            _txProvider.Unregister(registration);

            _log.Debug("Leave Unregister");
        }

        public virtual void Handle_Dispatcher(IEnvelopeDispatcher dispatcher)
        {
            _log.Debug("Enter Handle_Dispatcher");

            try
            {
                EnvelopeContext context = new EnvelopeContext(EnvelopeContext.Directions.In, dispatcher.Envelope);

                // send the envelope through the inbound processing chain
                this.ProcessEnvelope(context, this.InboundChain.Sort(), () =>
                {
                    dispatcher.Dispatch(context.Envelope);
                    _log.Debug("Dispatched envelope");
                });
            }
            catch (Exception ex)
            {
                _log.Warn("Failed to dispatch envelope; raising EnvelopeFailed event");
                dispatcher.Fail(ex);
            }

            _log.Debug("Leave Handle_Dispatcher");
        }

        public void Dispose()
        {
            _log.Debug("Enter Dispose");
            _txProvider.Dispose();
            _log.Debug("Leave Dispose");
        }


        // recursive function that processes envelopes
        protected virtual void ProcessEnvelope(
            EnvelopeContext context, 
            IEnumerable<IEnvelopeProcessor> processorChain, 
            Action processingComplete)
        {
            // if the chain is null or empty, complete processing
            if ((null == processorChain) || (!processorChain.Any()))
            {
                processingComplete();
                return;
            }

            // grab the next procesor
            IEnvelopeProcessor nextProcessor = processorChain.First();

            // let it process the envelope and pass its "next" processor: a
            // method that recursively calls this function with the current
            // processor removed
            nextProcessor.ProcessEnvelope(context, () =>
            {
                this.ProcessEnvelope(context, processorChain.Skip(1), processingComplete);
            });
        }
    }




    public static class ChainExtensions
    {
        public static IEnumerable<IEnvelopeProcessor> Sort(this IDictionary<int, IEnvelopeProcessor> chain)
        {
            IEnumerable<IEnvelopeProcessor> sortedChain = new List<IEnvelopeProcessor>();

            if (null != chain)
            {
                sortedChain = chain
                    .OrderBy(kvp => kvp.Key)
                    .Select<KeyValuePair<int, IEnvelopeProcessor>, IEnvelopeProcessor>(kvp => kvp.Value);
            }

            return sortedChain;
        }
    }
}
